test/state/host.cpp

```diff
@@ -0,0 +1,369 @@
+        assert(m_state.get(msg.sender).balance >= value);
+        m_state.get(msg.sender).balance -= value;
+        recipient_acc.balance += value;
+    }
+
+    // Copy of the code. Revert will invalidate the account.
+    const auto code = code_acc != nullptr ? code_acc->code : bytes{};
+    return m_vm.execute(*this, m_rev, msg, code.data(), code.size());
+}
+
+evmc::Result Host::call(const evmc_message& orig_msg) noexcept
+{
+    const auto msg = prepare_message(orig_msg);
+    if (!msg.has_value())
+        return evmc::Result{EVMC_OUT_OF_GAS, orig_msg.gas};  // Light exception.
+
+    auto state_snapshot = m_state;
+    auto destructs_snapshot = m_destructs.size();
+    auto access_addresses_snapshot = m_accessed_addresses;
+    auto logs_snapshot = m_logs.size();
+
+    auto result = execute_message(*msg);
+
+    if (result.status_code != EVMC_SUCCESS)
+    {
+        static constexpr auto addr_03 = 0x03_address;
+        auto* const acc_03 = m_state.get_or_null(addr_03);
+        const auto is_03_touched = acc_03 != nullptr && acc_03->touched;
+
+        // Revert.
+        m_state = std::move(state_snapshot);
+        m_destructs.resize(destructs_snapshot);
+        m_accessed_addresses = std::move(access_addresses_snapshot);
+        m_logs.resize(logs_snapshot);
+
+        // The 0x03 quirk: the touch on this address is never reverted.
+        if (is_03_touched && m_rev >= EVMC_SPURIOUS_DRAGON)
+            m_state.get_or_create(addr_03).touched = true;
+    }
+    return result;
+}
+
+evmc_tx_context Host::get_tx_context() const noexcept
+{
+    // TODO: The effective gas price is already computed in transaction validation.
+    const auto priority_gas_price =
+        std::min(m_tx.max_priority_gas_price, m_tx.max_gas_price - m_block.base_fee);
+    const auto effective_gas_price = m_block.base_fee + priority_gas_price;
+
+    return evmc_tx_context{
+        intx::be::store<uint256be>(effective_gas_price),  // By EIP-1559.
+        m_tx.sender,
+        m_block.coinbase,
+        m_block.number,
+        m_block.timestamp,
+        m_block.gas_limit,
+        m_block.prev_randao,
+        0x01_bytes32,  // Chain ID is expected to be 1.
+        uint256be{m_block.base_fee},
+    };
+}
+
+bytes32 Host::get_block_hash(int64_t block_number) const noexcept
+{
+    (void)block_number;
+    // TODO: This is not properly implemented, but only single state test requires BLOCKHASH
+    //       and is fine with any value.
+    return {};
+}
+
+void Host::emit_log(const address& addr, const uint8_t* data, size_t data_size,
+    const bytes32 topics[], size_t topics_count) noexcept
+{
+    m_logs.push_back({addr, {data, data_size}, {topics, topics + topics_count}});
+}
+
+evmc_access_status Host::access_account(const address& addr) noexcept
+{
+    // TODO: Predefined warm addresses can be applied to the state cache before execution.
+
+    // Transaction {sender,to} are always warm.
+    if (addr == m_tx.to)
```

Comment:
Why not add all these and access list to `m_accessed_addresses` in `Host` constructor?

Reply:
On todo list. I have it done already so I will include the change.

Reply:
Done, but a lot of new code landed.